wfl/fit/ref_error.py

```diff
@@ -1,213 +1,222 @@
+            generic_calc_kwargs = { "autopara_info": calc_autopara_info }
+        if calc_outputs is None:
+            calc_outputs = OutputSpec()
+        inputs = generic_calc(inputs, calc_outputs, calc_property_prefix, properties=calculator_properties, output_prefix="ref_error_calc_",
+                              **generic_calc_kwargs)
+        calc_property_prefix = "ref_error_calc_"
+
+    # clean up category_keys
+    if isinstance(category_keys, str):
+        category_keys = [category_keys]
+    elif category_keys is None:
+        category_keys = []
+
+    def _reshape_normalize(diff, at):
+        if per_component:
+            # one long vector
+            diff = diff.reshape((-1))
+        elif prop in atom_properties:
+            # flatten any vector/matrix dimensions so norm below is correct
+            diff = diff.reshape((len(at), -1))
+
+        if per_atom:
+            diff /= len(at)
+
+        return diff
+
+    # compute diffs and store in all_diffs, and weights in all_weights
+    all_diffs = {}
+    all_parity = { "ref": {}, "calc": {} }
+    all_weights = {}
+    for at in inputs:
+        # turn category keys into a single string for dict key
+        at_category = " / ".join([str(at.info.get(k)) for k in category_keys])
+        weight = at.info.get(weight_property, 1.0)
+
+        if len(set(config_properties).intersection(set(atom_properties))) > 0:
+            raise ValueError(f"Property {set(config_properties).intersection(set(atom_properties))} "
+                             "appears in both config_properties and atom_properties")
+
+        for prop in config_properties + atom_properties:
+            prop_use = prop
+
+            # parse (and remove) "/..." suffixes
+            per_atom = re.search(r"/atom\b", prop_use)
+            prop_use = re.sub(r"/atom\b", "", prop_use)
+            per_component = re.search(r"/comp\b", prop)
+            prop_use = re.sub(r"/comp\b", "", prop_use)
+            by_species = re.search(r"/Z\b", prop_use)
+            prop_use = re.sub(r"/Z\b", "", prop_use)
+
+            # possibly reconstruct virial later
+            virial_from_stress = False
+            if prop_use == "virial":
+                prop_use = "stress"
+                virial_from_stress = True
+
+            # select dict and check for inconsistencies
+            if prop in config_properties:
+                if by_species:
+                    raise ValueError("/Z only possible in atom_properties")
+                data = at.info
+            else: # atom_properties
+                if per_atom:
+                    raise ValueError("/atom only possible in config_properties")
+                data = at.arrays
+
+            # grab data
+            ref_quant = data.get(ref_property_prefix + prop_use)
+            calc_quant = data.get(calc_property_prefix + prop_use)
+            # skip if data is missing
+            if ref_quant is None or calc_quant is None:
```

Comment:
I'm definitely ok with a flag controlling this, but I'm not sure how I feel about the default being restrictive. I think it'll be very common to generally want virial errors, but to pass fitting databases with isolated atoms and dimers that have no stress.  I've implemented this in 1c95b7e, but I'm open to being convinced to flip the default.

Reply:
Is there a way to easily detect at least isolated atoms without relying on `atoms.info`?

Reply:
> Is there a way to easily detect at least isolated atoms without relying on `atoms.info`?

Not reliably. Specifically for virial/stress you can check for `atoms.pbc`, which is False for nonperiodic systems, so we could special case that - it would catch isolate atom and my dimers.

Reply:
ok in that case let's just have a warning and revert `1c95b7e`? Sorry for the back and forth.

Reply:
Fixed as of 9d52f0c.

Reply:
Feel free to un-resolve if there's anything else.